subcommand/cert-manager/command.go

```diff
@@ -0,0 +1,229 @@
+	c.once.Do(c.init)
+	if err := c.flagSet.Parse(args); err != nil {
+		certGenLog.Error(err, "parsing flagSet")
+		return 1
+	}
+	if len(c.flagSet.Args()) > 0 {
+		certGenLog.Error(errors.New("should have no non-flag arguments"), "invalid arguments")
+		return 1
+	}
+
+	if c.clientset == nil {
+		config, err := rest.InClusterConfig()
+		if err != nil {
+			certGenLog.Error(err, "Error loading in-cluster K8S config")
+			return 1
+		}
+		c.clientset, err = kubernetes.NewForConfig(config)
+		if err != nil {
+			certGenLog.Error(err, "Error creating K8S client")
+			return 1
+		}
+	}
+
+	var certSource cert.Source = &cert.GenSource{
+		Name:   "Consul Webhook Certificates",
+		Hosts:  strings.Split(c.flagAutoHosts, ","),
+		Expiry: 3 * time.Minute,
+	}
+	if c.flagCertFile != "" {
+		certSource = &cert.DiskSource{
+			CertPath: c.flagCertFile,
+			KeyPath:  c.flagKeyFile,
+		}
+	}
+
+	// Create the certificate notifier so we can update for certificates,
+	// then start all the background routines for updating certificates.
+	certCh := make(chan cert.Bundle)
+	certNotify := &cert.Notify{Ch: certCh, Source: certSource}
+	defer certNotify.Stop()
+	go certNotify.Start(context.Background())
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	c.certWatcher(ctx, certCh, c.clientset, certGenLog)
+	return 0
+}
+
+func (c *Command) certWatcher(ctx context.Context, ch <-chan cert.Bundle, clientset kubernetes.Interface, log logr.Logger) {
+	var bundle cert.Bundle
+	for {
+		select {
+		case bundle = <-ch:
+			log.Info("Updated certificate bundle received. Updating webhook certs.")
+			// Bundle is updated, set it up
+
+		case <-time.After(30 * time.Minute):
+			// This forces the mutating ctrlWebhook config to remain updated
+			// fairly quickly. This is a jank way to do this and we should
+			// look to improve it in the future. Since we use Patch requests
+			// it is pretty cheap to do, though.
+
+		case <-ctx.Done():
+			// Quit
+			return
+		}
+
+		certSecret, err := clientset.CoreV1().Secrets(c.flagSecretNamespace).Get(ctx, c.flagSecretName, metav1.GetOptions{})
+		if err != nil && k8serrors.IsNotFound(err) {
+			secret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      c.flagSecretName,
+					Namespace: c.flagSecretNamespace,
+				},
+				Data: map[string][]byte{
+					corev1.TLSCertKey:       bundle.Cert,
+					corev1.TLSPrivateKeyKey: bundle.Key,
+				},
+				Type: corev1.SecretTypeTLS,
+			}
+			_, err := clientset.CoreV1().Secrets(c.flagSecretNamespace).Create(ctx, secret, metav1.CreateOptions{})
+			if err != nil {
+				log.Error(err, "Error writing secret to API server")
+				continue
+			}
+			err = c.updateWebhookConfig(bundle, clientset, ctx, log)
+			if err != nil {
+				continue
+			}
+		} else if err != nil {
+			log.Error(err, "Error getting secret for API server")
+			continue
+		}
+
+		if bytes.Equal(certSecret.Data[corev1.TLSCertKey], bundle.Cert) {
```

Comment:
Ensure we only update the certs when they have been updated to prevent constantly rewriting the secret